### PR TITLE
Add documentation around FRAME Origin

### PR DESCRIPTION
**Pull request by kianenigma**
Does the following: 

- Add a reference doc page named `frame_runtime_types`, which explains what types like `RuntimeOrigin`, `RuntimeCall` etc are. 
- On top of it, it adds a reference doc page called `frame_origin` which explains a few important patterns that we use around origins
- And finally brushes up `#[frame::origin]` docs. 
- Updates the theme, sidebar and favicon to look like: 

Screenshot 2024-02-20 at 12 16 00

All of this was inspired by https://substrate.stackexchange.com/questions/10992/how-do-you-find-the-public-key-for-the-medium-spender-track-origin/10993

closes https://github.com/paritytech/polkadot-sdk-docs/issues/45
closes https://github.com/paritytech/polkadot-sdk-docs/issues/43
contributes / overlaps with https://github.com/paritytech/polkadot-sdk/pull/2638 cc @liamaharon 
deprecation companion: https://github.com/substrate-developer-hub/substrate-docs/pull/2131
pba-content companion: https://github.com/Polkadot-Blockchain-Academy/pba-content/pull/977

### CODE REVIEW

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 1]
```diff
@@ -1,14 +1,265 @@
...
+//! **ORIGIN IS NOT AN ACCOUNT ID**.
+//!
+//! FRAME's origin abstractions allow you to convey meanings far beyond just an account-id being the
+//! caller of an extrinsic. Nonetheless, an account-id having signed an extrinsic is one of the
+//! meanings that an origin can convey. This is the commonly used [`frame_system::ensure_signed`],
+//! where the return value happens to be an account-id.
+//!
+//! Instead, let's establish the following as the correct definition of an origin:
+//!
+//! > The origin type represent the privilege level of the caller of an extrinsic.
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_origin.rs:14]:
```suggestion
//! > The origin type represents the privilege level of the caller of an extrinsic.
```

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 2]
```diff
@@ -1,14 +1,265 @@
...
+//! For example, [`pallet_collective`] defines [`pallet_collective::EnsureMember`] and
+//! [`pallet_collective::EnsureProportionMoreThan`] and many more, which is exactly what we alluded
+//! to earlier in this document.
+//!
+//! Make sure to check the full list of [implementors of
+//! `EnsureOrigin`](frame::traits::EnsureOrigin#implementors) for more inspiration.
+//!
+//! ## Obtaining Abstract Origins
+//!
+//! So far we have learned that FRAME pallet's can assert on custom and abstract origin types,
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_origin.rs:114]:
```suggestion
//! So far we have learned that FRAME pallets can assert on custom and abstract origin types,
```

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 3]
```diff
@@ -1,14 +1,265 @@
...
+//! whether they are defined within the pallet or not. But how can we obtain these abstract origins?
+//!
+//! > All extrinsics that come from outer world can generally only obtained either `signed` or
+//! > `none` origin
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_origin.rs:118]:
```suggestion
//! > All extrinsics that come from the outer world can generally only be obtained as either `signed` or
//! > `none` origin
```

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 4]
```diff
@@ -1,14 +1,265 @@
...
+//!
+//! Generally, these abstract origins asd only obtained within the runtime, when a call is
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_origin.rs:120]:
```suggestion
//! Generally, these abstract origins are only obtained within the runtime, when a call is
```

File: docs/sdk/src/reference_docs/frame_runtime_types.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+//! > [#polkadot-sdk#1378](https://github.com/paritytech/polkadot-sdk/pull/1378) is meant to
+//! > significantly improve this. Exploring the rust-docs of a runtime, such as [`runtime`] which is
+//! > defined in this module is as of now the best way to learn about these types.
+//!
+//! ## Composite Enums
+//!
+//! Many types within a FRAME runtime follow the following structure:
+//!
+//! * Each individual pallet defines a type, for example `Foo`.
+//! * At thr runtime level, these types are amalgamated into a single type, for example
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_runtime_types.rs:17]:
```suggestion
//! * At the runtime level, these types are amalgamated into a single type, for example
```

File: docs/sdk/src/reference_docs/frame_runtime_types.rs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+//!   `RuntimeFoo`.
+//!
+//! As the names suggest, all composite enums in a FRAME runtime start their name with `Runtime`.
+//! For example, `RuntimeCall` is a representation of the most high level `Call`-able type in the
+//! runtime.
+//!
+//! Composite enums are generally convertible to their individual parts as such:
+#![doc = simple_mermaid::mermaid!("../../../mermaid/outer_runtime_types.mmd")]
+//!
+//! In that one can always convert from the inner type into the outer type, but not visa versa. This
```
Comment (DrW3RK) [inline docs/sdk/src/reference_docs/frame_runtime_types.rs:27]:
```suggestion
//! In that one can always convert from the inner type into the outer type, but not vice versa. This
```

**Review by DrW3RK — COMMENTED** [review]
Thanks for the write-up on origins!

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 7]
```diff
@@ -1,14 +1,263 @@
...
+		AllNftHolders,
+		/// If all validators have agreed upon this.
+		ValidatorSet,
+	}
+
+	#[docify::export(custom_origin_usage)]
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		pub fn only_validators(
+			// we use the full syntax of `OriginFor` here, but it's not necessary.
```
Comment (skunert) [inline docs/sdk/src/reference_docs/frame_origin.rs:182]: What does "full syntax" mean here? Leftover comment and was previously using full crate path?

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 8]
```diff
@@ -1,14 +1,263 @@
...
+			origin: OriginFor<T>,
+		) -> DispatchResult {
+			// first, we convert from `<T as frame_system>::RuntimeOrigin` to `<T as
```
Comment (skunert) [inline docs/sdk/src/reference_docs/frame_origin.rs:185]:
```suggestion
			// first, we convert from `<T as frame_system::Config>::RuntimeOrigin` to `<T as
```

File: docs/sdk/src/reference_docs/frame_runtime_types.rs [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+//! passed to `frame_system`.
+#![doc = docify::embed!("./src/reference_docs/frame_runtime_types.rs", pallet_with_specific_runtime_call_impl)]
+//!
+//! > In other words, the degree of specificity that [`frame_system::Config::RuntimeCall`] has is
+//! > not enough for the pallet to work with. Therefore, the pallet has to define its own associated
+//! > type.
+//!
+//! Another way to look at this is:
+//!
+//! `pallet_with_specific_runtime_call::Config::RuntimeCall` are two different representations of
```
Comment (skunert) [inline docs/sdk/src/reference_docs/frame_runtime_types.rs:97]: Something is missing here I think, what are two differen representations?

File: docs/sdk/src/reference_docs/frame_runtime_types.rs [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+//! the same concrete type that is only know at when the runtime is being constructed.
```
Comment (skunert) [inline docs/sdk/src/reference_docs/frame_runtime_types.rs:98]:
```suggestion
//! the same concrete type that is only known when the runtime is being constructed.
```

**Review by skunert — COMMENTED** [review]
Nice docs! 

When I run locally it seems like the theming is inconsistent. On some pages I have the default theme and it switches randomly to a darker theme (browser: firefox). Not sure if this is an issue on my end :thinking: 
Screenshot from 2024-02-22 10-44-41

vs

Screenshot from 2024-02-22 10-44-33

File: docs/sdk/src/reference_docs/frame_origin.rs (hunk 7 above)
Comment (kianenigma) [reply docs/sdk/src/reference_docs/frame_origin.rs:182]: yeah old comment.

Comment (kianenigma) [inline docs/sdk/src/reference_docs/frame_origin.rs:182]:
```suggestion
```

File: docs/sdk/src/reference_docs/mod.rs [hunk 11]
```diff
@@ -100,3 +100,5 @@ pub mod frame_runtime_migration;
 /// light-node-first out of the box.
 // TODO: @jsdw @josepot https://github.com/paritytech/polkadot-sdk-docs/issues/68
 pub mod light_nodes;
+
+// pub mod opinionated_pallets;
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/mod.rs:104]:
```suggestion
```

**Review by alvicsam — APPROVED** [review]

**Review by skunert — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5323568

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325131

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 12]
```diff
@@ -1,14 +1,260 @@
...
+//! Make sure to check the full list of [implementors of
+//! `EnsureOrigin`](frame::traits::EnsureOrigin#implementors) for more inspiration.
+//!
+//! ## Obtaining Abstract Origins
+//!
+//! So far we have learned that FRAME pallets can assert on custom and abstract origin types,
+//! whether they are defined within the pallet or not. But how can we obtain these abstract origins?
+//!
+//! > All extrinsics that come from the outer world can generally only be obtained as either
+//! > `signed` or `none` origin
```
Comment (gpestana) [inline docs/sdk/src/reference_docs/frame_origin.rs:118]:
```suggestion
//! > `signed` or `none` origin.
```

**Review by gpestana — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5337812

File: docs/sdk/src/reference_docs/mod.rs [hunk 13]
```diff
@@ -44,15 +44,15 @@ pub mod extrinsic_encoding;
 pub mod signed_extensions;
 
 /// Learn about *"Origin"* A topic in FRAME that enables complex account abstractions to be built.
```
Comment (liamaharon) [inline docs/sdk/src/reference_docs/mod.rs:46]:
```suggestion
/// Learn about *Origins*, a topic in FRAME that enables complex account abstractions to be built.
```

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 14]
```diff
@@ -1,14 +1,260 @@
...
-//!   Protocol Berg 2023
-//! - system's raw origin, how it is amalgamated with other origins into one type
-//! [`frame_composite_enums`]
-//! - signed origin
-//! - unsigned origin, link to [`fee_less_runtime`]
-//! - Root origin, how no one can obtain it.
-//! - Abstract origin: how FRAME allows you to express "origin is 2/3 of the this body or 1/2 of
-//!   that body or half of the token holders".
-//! - `type CustomOrigin: EnsureOrigin<_>` in pallets.
+//! let's start by eliminating a wrong assumption that is a common mistake:
```
Comment (liamaharon) [inline docs/sdk/src/reference_docs/frame_origin.rs:3]:
```suggestion
//! Let's start by clarifying a common wrong assumption about Origin:
```

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 15]
```diff
@@ -1,14 +1,260 @@
...
+//!
+//! That is, an extrinsic, through checking the origin, can *express what privilege level it wishes
+//! to impose on the caller of the extrinsic*. One of those checks can be as simple as "*any account
+//! that has signed a statement can pass*".
+//!
+//! But the origin system can also express more abstract and complicated privilege levels. For
+//! example:
+//!
+//! * If the majority of token holders agreed upon this. This is more or less what the
+//!   [`pallet_democracy`] does under the hood ([reference](https://github.com/paritytech/polkadot-sdk/blob/edd95b3749754d2ed0c5738588e872c87be91624/substrate/frame/democracy/src/lib.rs#L1603-L1633)).
```
Comment (liamaharon) [inline docs/sdk/src/reference_docs/frame_origin.rs:24]: Is this an example of using the Origin abstraction? Seems just like regular logic

File: docs/sdk/src/reference_docs/frame_origin.rs [hunk 16]
```diff
@@ -1,14 +1,260 @@
...
+//! First, let's look at where the `origin` type is encountered in a typical pallet. The `origin:
+//! OriginFor<T>` has to be the first argument of any given callable extrinsic in FRAME:
+#![doc = docify::embed!("./src/reference_docs/frame_origin.rs", call_simple)]
+//!
+//! Typically, the code of an extrinsic starts with an origin check, such as
+//! [`frame_system::ensure_signed`].
+//!
+//! Note that [`OriginFor`](frame_system::pallet_prelude::OriginFor) is merely a shorthand for
+//! [`frame_system::Config::RuntimeOrigin`]. Given the name prefix `Runtime`, we can learn that
+//! `RuntimeOrigin` is similar to `RuntimeCall` and others, a runtime composite  that is amalgamated
```
Comment (liamaharon) [inline docs/sdk/src/reference_docs/frame_origin.rs:43]:
```suggestion
//! `RuntimeOrigin` is similar to `RuntimeCall` and others, a runtime composite enum that is amalgamated
```

**Review by liamaharon — APPROVED** [review]
Some comments but overall doc looks good. Also, theme looks great 👍🏼

File: docs/sdk/src/reference_docs/frame_origin.rs (hunk 15 above)
Comment (kianenigma) [reply docs/sdk/src/reference_docs/frame_origin.rs:24]: It is meant to showcase that pallets can dispatch internally and set any origin. And, that an origin like root can only be acquired in this way. 

I will re-read it one more time.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5345545

**Review by muharem — APPROVED** [review]
Thank you!

It would be also very helpful to have more context for `OriginTrait` docs, specially methods `add_filter`, `reset_filter`, `filter_call`.

**Comment by kianenigma** [comment]
> Thank you!
> 
> It would be also very helpful to have more context for `OriginTrait` docs, specially methods `add_filter`, `reset_filter`, `filter_call`.

This is a great suggestion! will either do here or in a follow-up.

**Pull request merged**